feat!: Account linking (#670)
supertokens/supertokens-node: CHANGELOG.md

## [unreleased]

## [16.0.0] - 2023-08-XX

### Breaking changes

> 

Reviewer: Need to talk about higher level changes first:
- New concept of recipeUserId vs primary user ID you can link to docs here: supertokens.com/docs/thirdpartyemailpassword/user-object#primary-vs-recipe-user-id
- Introducing account linking functionality.

> -   Now only supporting CDI 4.0

Reviewer: ```suggestion
-   Now only supporting CDI 4.0. Compatible with core version >= 7.0
```

-   Now supporting FDI 1.18

> -   removed the recipe specific `User` type, now all functions are using the new generic `User` type

Reviewer: Also link to this page: supertokens.com/docs/thirdpartyemailpassword/user-object

> -   The `fetchValue` callback of claims now take a new `recipeUserId` param

Author: This affects all recipes that have claims (I'll add an exact list)

-   Now ignoring protected props in the payload in `createNewSession` and `createNewSessionWithoutRequestResponse`
-   `createdNewUser` has been renamed to `createdNewRecipeUser` in sign up related APIs and functions

-   EmailPassword:
    -   removed `getUserById`, `getUserByEmail`

>     -   added `consumePasswordResetToken`

Reviewer: ```suggestion
    -   added `consumePasswordResetToken`. This function allows the consumption of the reset password token without changing the password. It will return OK if the token was valid.
```

>     -   added an overrideable `createNewRecipeUser` function that is called during sign up and in the “invitation link” flow

Reviewer: ```suggestion
    -   added an overrideable `createNewRecipeUser` function that is called during sign up and password reset flow (in case a new email password user is being created on the fly). This is mostly for internal use.
```

    -   `recipeUserId` is added to the input of `getContent` of the email delivery config
    -   `email` was added to the input of `createResetPasswordToken` , `sendResetPasswordEmail`, `createResetPasswordLink`
    -   `updateEmailOrPassword` :
        -   now takes `recipeUserId` instead of `userId`
        -   can return the new `EMAIL_CHANGE_NOT_ALLOWED_ERROR` status
    -   `signIn`:
        -   returns `recipeUserId`
    -   `signUp`:

>         -   returns `recipeUserId`

Reviewer: - these return the whole new user object now right? Not just recipeUserId.
- Also, these functions now attempt to link accounts as well

Author: - Yup, but they already did that, only the type change (which is mentioned elsewhere)
- They do, but only if you turn on account-linking which is a whole separate docs. I'll add a few lines about that pointing to the right docs.

    -   `signInPOST`:
        -   can return status `SIGN_IN_NOT_ALLOWED`
    -   `signUpPOST`:
        -   can return status `SIGN_UP_NOT_ALLOWED`
    -   `generatePasswordResetTokenPOST`:
        -   can now return `PASSWORD_RESET_NOT_ALLOWED`
    -   `passwordResetPOST`:
        -   now returns the `user` and the `email` whose password was reset

>         -   can now return `PASSWORD_POLICY_VIOLATED_ERROR`

Reviewer: this was done before as well. Maybe you meant to use another status code?

Author: No, there is a related change in `lib/ts/recipe/emailpassword/types.ts` maybe it was a typing issue before.

-   Updated compilation target of typescript to ES2017 to make debugging easier.
-   Added account-linking recipe

### Migration guide

#### New User structure

We've added a generic `User` class instead of the old recipe specific ones. The mapping of old props to new in case you are not using account-linking:

-   `user.id` stays `user.id` (or `user.loginMethods[0].recipeUserId` in case you need `RecipeUserId`)
-   `user.email` becomes `user.emails[0]`
-   `user.phoneNumber` becomes `user.phoneNumbers[0]`
-   `user.thirdParty` becomes `user.thirdParty[0]`
-   `user.timeJoined` is still `user.timeJoined`
-   `user.tenantIds` is still `user.tenantIds`

#### RecipeUserId

Some functions now require you to pass a `RecipeUserId` instead of a string user id. If you are using our auth recipes, you can find the recipeUserId as: `user.loginMethods[0].recipeUserId` (you'll need to worry about selecting the right login method after enabling account linking). Alternatively, if you already have a string user id you can convert it to a `RecipeUserId` using `supertokens.convertToRecipeUserId(userIdString)`

> #### Checking if a new primary user was created

Reviewer: ```suggestion
#### Checking if a user signed up or signed in
```

> -   You can check if a new primary user created by `ThirdParty.manuallyCreateOrUpdateUser`, `signInUp`, `signInUpPOST`, `Passwordless.consumeCode` or `consumeCodePOST` (and their combination recipe counterparts):

Reviewer: ```suggestion
-   In the emailpassword sign up / passwordless consumeCode / social login signinup APIs, you can check if a user signed up by:
```

```
    // Here res refers to the result the function/api functions mentioned above.

>     const isNewPrimaryUser = res.createdNewRecipeUser && res.user.loginMethod.length === 1;

Reviewer: ```suggestion
    const isNewUser = res.createdNewRecipeUser && res.user.loginMethod.length === 1;
```

```

-   You can check if a new primary user was created by `EmailPassword.signUp`, `signUpPOST` or `createNewRecipeUser` (and their ThirdParyEmailPassword counterparts) by:

```
    const isNewPrimaryUser = res.user.loginMethod.length === 1;

> ```

Reviewer: huh? Not sure why this is needed. Did you mean that you can check if a new user signed in?

Author: No, this is different from the above, since EP signup doesn't have `createdNewRecipeUser`

#### Changing user emails

-   We recommend that you check if the email change of a user is allowed, before calling the update function
    -   Check [here](https://supertokens.com/docs/thirdpartyemailpassword/common-customizations/change-email-post-login) for more information

```
import {isEmailChangeAllowed} from "supertokens-node/recipe/accountlinking";
/// ...
app.post("/change-email", verifySession(), async (req: SessionRequest, res: express.Response) => {
    let session = req.session!;
    let email = req.body.email;

    // ...
    if (!(await isEmailChangeAllowed(session.getRecipeUserId(), email, false))) {
        // this can come here if you have enabled the account linking feature, and
        // if there is a security risk in changing this user's email.
    }

    // Update the email
    let resp = await ThirdPartyEmailPassword.updateEmailOrPassword({
        recipeUserId: session.getRecipeUserId(),
        email: email,
    });
    // ...
});
```

> 

Reviewer: Need to have a section on frontend / core migration needed - that these are the versions of the frontend and core.

Author: The supported FE version didn't change if you are not using account linking, and the core is mentioned above. I can add a section/link here pointing to the account linking recipe docs.
